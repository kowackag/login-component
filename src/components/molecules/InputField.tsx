import { Path, UseFormRegister } from "react-hook-form";

interface IFormValues {
	email: string;
	password: number;
	saved: boolean;
}
type FormFieldTypes = {
	name: Path<IFormValues>;
	label: string;
	type?: "text" | "password" | "email";
	register: UseFormRegister<{ email: string; password: string; saved: boolean }>;
	error: string | undefined;
};

export const InputField = ({ name, label, type, register, error }: FormFieldTypes) => {
	return (
		<div className="relative py-3">
			<label className="font-primaryMedium inline-block" htmlFor={name}>
				{label}
			</label>
			<input
				id={name}
				type={type}
				{...register(name)}
				className="w-full rounded-md border-2 border-neutral-300 bg-transparent px-2 leading-9 outline-none"
			/>
			{error ? (
				<p className="font-primaryMedium absolute right-0 text-xs italic text-red-600">{error}</p>
			) : null}
		</div>
	);
};
