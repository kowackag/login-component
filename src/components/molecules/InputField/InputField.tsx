import { ChangeEventHandler, FocusEventHandler, Ref } from "react";
import { Path } from "react-hook-form";

import { ErrorMessage } from "@/components/atoms/Error/Error";

export interface IFormValues {
	email: string;
	password: number;
	toSave: boolean;
}

export type InputFieldTypes = {
	name: Path<IFormValues>;
	label: string;
	type?: HTMLInputElement["type"];
	onChange: ChangeEventHandler;
	onBlur: FocusEventHandler<HTMLInputElement>;
	inputRef: Ref<HTMLInputElement>;
	error: string | undefined;
};

export const InputField = ({
	name,
	label,
	type = "text",
	onChange,
	onBlur,
	inputRef,
	error,
}: InputFieldTypes) => {
	return (
		<div className="relative py-3">
			<label className="inline-block font-primaryMedium" htmlFor={name}>
				{label}
			</label>
			<input
				id={name}
				name={name}
				type={type}
				onChange={onChange}
				onBlur={onBlur}
				ref={inputRef}
				data-testid={`${name}-input`}
				className="w-full rounded-md border-2 border-neutral-300 bg-transparent px-2 leading-9 outline-none"
			/>
			{error ? <ErrorMessage message={error} /> : null}
		</div>
	);
};
