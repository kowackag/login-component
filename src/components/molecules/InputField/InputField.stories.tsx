import type { Meta, StoryObj } from "@storybook/react";

import { InputField } from "./InputField";

const meta: Meta<typeof InputField> = {
	component: InputField,
	decorators: [
		(Story) => (
			<div style={{ width: "360px" }}>
				<Story />
			</div>
		),
	],
};

export default meta;

type Story = StoryObj<typeof InputField>;

export const Default: Story = {
	args: {
		label: "Login",
		error: "it is required",
	},
};
