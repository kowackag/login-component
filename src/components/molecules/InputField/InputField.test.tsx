import { describe, expect, it } from "vitest";
import { fireEvent, render, screen } from "@testing-library/react";
import { InputField } from "./InputField";

describe("InputField", () => {
	it("emailInput", () => {
		const { container } = render(<InputField name="email" label="email" error=""  />);

		const emailInput = container.querySelector("#email") as HTMLInputElement;
		fireEvent.change(emailInput, { target: { value: "email@google.com" } });
		expect(emailInput.value).toBe("email@google.com");
	});
	it("passwordInput", () => {
		const { container } = render(
			<InputField name="password" label="password" error="" type="password" />,
		);
		const passwordInput = container.querySelector("#password") as HTMLInputElement;
		fireEvent.change(passwordInput, { target: { value: "password" } });

		expect(passwordInput.value).toHaveDisplayValue;
	});
	it("input error is shown", () => {
		render(<InputField name="email" label="email" error="some error" />);
		const errorEl = screen.getByTestId("error");
		expect(errorEl).toBeVisible();
            
		//expect(errorEl.children).toBe("some error");
	});
	it("input error is not rendered", () => {
		const { container } = render(<InputField name="email" label="email" error="" />);
		const emailInput = container.querySelector("#email") as HTMLInputElement;
		expect(emailInput.nextElementSibling).toBe(null);
	});
});
