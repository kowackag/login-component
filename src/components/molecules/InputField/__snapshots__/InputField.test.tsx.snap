// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InputField > renders correctly 1`] = `
<div>
  <div
    class="relative py-3"
  >
    <label
      class="inline-block font-primaryMedium"
      for="email"
    >
      email
    </label>
    <input
      class="w-full rounded-md border-2 border-neutral-300 bg-transparent px-2 leading-9 outline-none"
      data-testid="email-input"
      id="email"
      name="email"
      type="text"
    />
  </div>
</div>
`;
